src/Identity/Extensions.Stores/src/UserStoreBase.cs

```diff
@@ -920,7 +920,20 @@ public virtual async Task<int> CountCodesAsync(TUser user, CancellationToken can
         var mergedCodes = await GetTokenAsync(user, InternalLoginProvider, RecoveryCodeTokenName, cancellationToken).ConfigureAwait(false) ?? "";
         if (mergedCodes.Length > 0)
         {
-            return mergedCodes.Split(';').Length;
+            // non-allocating version of mergedCodes.Split(';').Length
+            var count = 1;
+            var index = 0;
+            while (index < mergedCodes.Length)
+            {
+                var semiColonIndex = mergedCodes.IndexOf(';', index);
+                if (semiColonIndex < 0)
+                {
+                    break;
+                }
+                count++;
+                index = semiColonIndex + 1;
+            }
+            return count;
```

Comment:
This one makes me a little sad. Feels like something missing API?

cc @stephentoub

Reply:
https://github.com/dotnet/runtime/issues/59466

Reply:
finally got around to do it. PR here: dotnet/runtime#80662